File: kernel/internal/tools/fs_tools.go
package tools

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"

	"agentos/kernel/internal/resource"
	"agentos/kernel/internal/sandbox"
)

// FSReadTool 读文本文件。
type FSReadTool struct{}

func (FSReadTool) Name() string { return "fs_read" }

func (FSReadTool) Schema() json.RawMessage {
	return json.RawMessage(`{
		"name": "fs_read",
		"description": "Read a text file. Path is relative to the workspace root.",
		"parameters": {"type":"object","properties":{"path":{"type":"string"}},"required":["path"]}
	}`)
}

func (FSReadTool) PermissionKey(params map[string]any) resource.Resource {
	path, _ := params["path"].(string)
	return resource.Resource{Type: "path", ID: path}
}

func (FSReadTool) Execute(ctx context.Context, params map[string]any) (Result, error) {
	path, _ := params["path"].(string)
	safe, err := sandbox.Resolve(path)
	if err != nil {
		return Result{}, err
	}
	data, err := os.ReadFile(safe)
	if err != nil {
		return Result{}, err
	}
	return Result{Data: map[string]any{"content": string(data)}}, nil
}

// FSWriteTool 写文本文件，原子写（.tmp → rename）防半成品。
type FSWriteTool struct{}

func (FSWriteTool) Name() string { return "fs_write" }

func (FSWriteTool) Schema() json.RawMessage {
	return json.RawMessage(`{
		"name": "fs_write",
		"description": "Write text content to a file (creates parent dirs).",
		"parameters": {"type":"object","properties":{"path":{"type":"string"},"content":{"type":"string"}},"required":["path","content"]}
	}`)
}

func (FSWriteTool) PermissionKey(params map[string]any) resource.Resource {
	path, _ := params["path"].(string)
	return resource.Resource{Type: "path", ID: path}
}

func (FSWriteTool) Execute(ctx context.Context, params map[string]any) (Result, error) {
	path, _ := params["path"].(string)
	content, _ := params["content"].(string)
	safe, err := sandbox.Resolve(path)
	if err != nil {
		return Result{}, err
	}
	if err := os.MkdirAll(filepath.Dir(safe), 0755); err != nil {
		return Result{}, err
	}
	// 原子写：先写 .tmp，再 rename。崩溃时要么旧文件要么新文件，无半成品。
	tmp := safe + ".tmp"
	if err := os.WriteFile(tmp, []byte(content), 0644); err != nil {
		return Result{}, err
	}
	if err := os.Rename(tmp, safe); err != nil {
		_ = os.Remove(tmp)
		return Result{}, err
	}
	return Result{Data: map[string]any{"bytes_written": len(content)}}, nil
}

// FSListTool 列目录条目。
type FSListTool struct{}

func (FSListTool) Name() string { return "fs_list" }

func (FSListTool) Schema() json.RawMessage {
	return json.RawMessage(`{
		"name": "fs_list",
		"description": "List file names in a directory.",
		"parameters": {"type":"object","properties":{"path":{"type":"string"}},"required":["path"]}
	}`)
}

func (FSListTool) PermissionKey(params map[string]any) resource.Resource {
	path, _ := params["path"].(string)
	return resource.Resource{Type: "path", ID: path}
}

func (FSListTool) Execute(ctx context.Context, params map[string]any) (Result, error) {
	path, _ := params["path"].(string)
	safe, err := sandbox.Resolve(path)
	if err != nil {
		return Result{}, err
	}
	entries, err := os.ReadDir(safe)
	if err != nil {
		return Result{}, err
	}
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		names = append(names, e.Name())
	}
	return Result{Data: map[string]any{"entries": names}}, nil
}

// 接口实现静态断言
var _ Tool = FSReadTool{}
var _ Tool = FSWriteTool{}
var _ Tool = FSListTool{}

File: cp/tools/fs_tools.py
import asyncio
import json
import os
from typing import Any, Dict

from cp.resource import Resource
from cp.sandbox.fs import resolve
from cp.tools.tool import ToolResult


class FSReadTool:
    def name(self) -> str:
        return "fs_read"

    def schema(self) -> str:
        return json.dumps({
            "name": "fs_read",
            "description": "Read a text file. Path is relative to the workspace root.",
            "parameters": {"type": "object",
                           "properties": {"path": {"type": "string"}},
                           "required": ["path"]},
        })

    def permission_key(self, params: Dict[str, Any]) -> Resource:
        return Resource(type="path", id=params.get("path", ""))

    async def execute(self, ctx: Any, params: Dict[str, Any]) -> ToolResult:
        safe = resolve(params.get("path", ""))
        content = await asyncio.to_thread(self._read_file, safe)
        return ToolResult(data={"content": content})

    def _read_file(self, path):
        with open(path, "r", encoding="utf-8") as f:
            return f.read()


class FSWriteTool:
    def name(self) -> str:
        return "fs_write"

    def schema(self) -> str:
        return json.dumps({
            "name": "fs_write",
            "description": "Write text content to a file (creates parent dirs).",
            "parameters": {"type": "object",
                           "properties": {"path": {"type": "string"},
                                          "content": {"type": "string"}},
                           "required": ["path", "content"]},
        })

    def permission_key(self, params: Dict[str, Any]) -> Resource:
        return Resource(type="path", id=params.get("path", ""))

    async def execute(self, ctx: Any, params: Dict[str, Any]) -> ToolResult:
        safe = resolve(params.get("path", ""))
        content = params.get("content", "")
        await asyncio.to_thread(self._write_file, safe, content)
        return ToolResult(data={"bytes_written": len(content)})

    def _write_file(self, safe, content):
        os.makedirs(os.path.dirname(safe), exist_ok=True)
        # 原子写:先写 .tmp,再 rename。崩溃时要么旧要么新,无半成品。
        tmp = safe + ".tmp"
        with open(tmp, "w", encoding="utf-8") as f:
            f.write(content)
        os.replace(tmp, safe)


class FSListTool:
    def name(self) -> str:
        return "fs_list"

    def schema(self) -> str:
        return json.dumps({
            "name": "fs_list",
            "description": "List file names in a directory.",
            "parameters": {"type": "object",
                           "properties": {"path": {"type": "string"}},
                           "required": ["path"]},
        })

    def permission_key(self, params: Dict[str, Any]) -> Resource:
        return Resource(type="path", id=params.get("path", ""))

    async def execute(self, ctx: Any, params: Dict[str, Any]) -> ToolResult:
        safe = resolve(params.get("path", ""))
        names = await asyncio.to_thread(self._list_dir, safe)
        return ToolResult(data={"entries": names})

    def _list_dir(self, path):
        return sorted(os.listdir(path))
